File: acquire/helpers/PSeq_Excite_PE.py
import numpy as np
import pypulseq as pp

from .PSeq_Base import PSeq_Base

class PSeq_Excite_PE(PSeq_Base):
    """
    Sequence component that plays traps on each axis (or fewer).

    This is made to be fairly flexible in designing the shortest possible time gradients,
    to capture a given starting moment or ending moment.  The intended use is at the end
    of a TR for refocusing/spoiling, but it can also be used for phase encoding or other uses.
    """

    def __init__(
        self,
        pparams,
        duration = 6e-3, thickness = 1e-3, 
                 tbw = 4, flip = 50, app = 0.4, 
                 fov = 200e-3, N_pe = 0,
                 max_slew = 100,
                 do_refocus = True,
                 *args, **kwargs
    ):
        """
        Construct.

        Parameters
        ----------

        """
        super().__init__(pparams)
        
        self.duration = duration
        self.do_refocus = do_refocus
        self.fov = fov
        self.N_pe = N_pe
        self.thickness = thickness
        self.tbw = tbw
        if flip > np.pi:  # Assume anything higher than pi is in degrees
            flip = flip * np.pi / 180
        self.flip = flip
        self.app = app
        
        if max_slew is not None:
            self.slew = self.pparams.system.gamma*max_slew
        else:
            self.slew = self.pparams.system.max_slew

        self.rfp, self.gss, self.gss_re = pp.make_sinc_pulse(flip_angle=self.flip, apodization=self.app, duration=self.duration, 
                                                system=self.pparams.system, time_bw_product=self.tbw, delay=self.pparams.system.rf_dead_time,
                                                slice_thickness=self.thickness, return_gz=True, max_slew = self.slew)
        
        self.refocus_time = 0
        if N_pe > 0:
            self.pe_areas = (np.arange(N_pe) - N_pe//2)/fov
            self.max_area = np.max(np.abs(self.pe_areas))
            pe_temp = pp.make_trapezoid(channel = self.pparams.channels[0], area=self.max_area, system=self.pparams.system)
            self.refocus_time = pp.calc_duration(pe_temp)
        if do_refocus:
            if pp.calc_duration(self.gss_re) >= self.refocus_time:
                self.refocus_time = pp.calc_duration(self.gss_re)
            else:
                self.gss_re = pp.make_trapezoid(channel = self.pparams.channels[2], duration = self.refocus_time, 
                                                area=self.gss_re.area, system=self.pparams.system)
                
        if N_pe > 0:
            self.pe_grad0 = pp.make_trapezoid(channel = self.pparams.channels[0], duration = self.refocus_time, 
                                                    area=self.max_area, system=self.pparams.system)        
            
            self.pe_grad1 = pp.make_trapezoid(channel = self.pparams.channels[1], duration = self.refocus_time, 
                                                    area=self.max_area, system=self.pparams.system)  
            
            self.amp0 = self.pe_grad0.amplitude
            self.amp1 = self.pe_grad1.amplitude

    def build_blocks(
        self,
        idx0=0,
        idx1=0,
        offset=0
    ):
        """Build all blocks to add to sequence.

        Parameters
        ----------
        idx0, idx1 : int
            Index for the 0th channel area.  Will be ignored if the areas given for this
            channel were not inputted with multiple values.

        Returns
        -------
        list of lists
            All blocks to add, outer list is for sequential blocks, inner blocks are all
            components to play within the block.
        """
        self.gss.channel = self.pparams.channels[2]
        self.rfp.freq_offset = self.gss.amplitude * offset
        
        all_blocks = [[self.rfp, self.gss]]
        
        grads_to_play = []
        
        if self.N_pe > 0:
            area0 = self.pe_areas[idx0]
            area1 = self.pe_areas[idx1]
            
            self.pe_grad0.channel = self.pparams.channels[0]
            self.pe_grad1.channel = self.pparams.channels[1]
            
            self.pe_grad0.amplitude = self.amp0 * area0/self.max_area
            self.pe_grad1.amplitude = self.amp1 * area1/self.max_area
            
            grads_to_play.append(self.pe_grad0)
            grads_to_play.append(self.pe_grad1)
            
            
        if self.do_refocus:
            self.gss_re.channel = self.pparams.channels[2]
            grads_to_play.append(self.gss_re)
            
        all_blocks.append(grads_to_play)
        return all_blocks

File: acquire_matlab/helpers/PSeq_Excite_PE.m
classdef PSeq_Excite_PE < PSeq_Base
    % Sequence component that plays slice excitation and optional phase encoding
    % 
    % This is made for GIRF measurements, so the FOV is assumed to be square.

    properties
        duration;
        thickness;
        tbw;
        flip;
        app;
        fov;
        N_pe;
        max_slew;
        do_refocus;
        do_prewind;

        slew;
        refocus_time;
        pe_areas;
        max_area;
        
        rfp;
        gss;
        gss_re;
        pe_grad1;
        pe_grad2;
        amp1;
        amp2;
    end

    methods
        function obj = PSeq_Excite_PE(pparams,varargin)

            obj = obj@PSeq_Base(pparams);

            p = inputParser;
            addParameter(p,'duration',6e-3);
            addParameter(p,'thickness',1e-3);
            addParameter(p,'tbw',4);
            addParameter(p,'flip',50);
            addParameter(p,'app',0.4);
            addParameter(p,'fov',200e-3);
            addParameter(p,'N_pe',0);
            addParameter(p,'max_slew',-1);
            addParameter(p,'do_refocus',true);
            addParameter(p,'do_prewind',true);

            parse(p, varargin{:});

            fields = fieldnames( p.Results );

            for n = 1:numel( fields )
	            obj.( fields{ n } ) = p.Results.( fields{ n } );
            end

            % ----------------------
            if obj.flip > pi
                obj.flip = obj.flip * pi / 180;
            end

            if obj.max_slew > 0
                obj.slew = obj.pparams.sys.gamma*maxSlew;
            else
                obj.slew = obj.pparams.sys.maxSlew;
            end


            % Get RF, slice select and refocus/prewinding gradient (assumes RF is coming out symmetric for prewinder)
            [obj.rfp, obj.gss, obj.gss_re] = mr.makeSincPulse(obj.flip, 'apodization', obj.app, 'duration', obj.duration, ...
                                                    'system', obj.pparams.sys, 'timeBwProduct', obj.tbw, 'delay', obj.pparams.sys.rfDeadTime, ...
                                                    'sliceThickness', obj.thickness, 'maxSlew', obj.slew, 'use', 'excitation');
            
            % Figure out the longest time needed between refocusing and phase encoding
            obj.refocus_time = 0;
            if obj.N_pe > 0
                obj.pe_areas = ((0:obj.N_pe-1) - floor(obj.N_pe/2))/obj.fov;
                obj.max_area = max(abs(obj.pe_areas));  % This gets used to rescale amplitude
                pe_temp = mr.makeTrapezoid(obj.pparams.channels{1}, 'area', obj.max_area, 'system', obj.pparams.sys);
                obj.refocus_time = mr.calcDuration(pe_temp);
            end
            if obj.do_refocus
                if mr.calcDuration(obj.gss_re) >= obj.refocus_time
                    obj.refocus_time = mr.calcDuration(obj.gss_re);
                else
                    % Remake gss_re to match longer phase encode time
                    obj.gss_re = mr.makeTrapezoid(obj.pparams.channels{3}, 'duration', obj.refocus_time, ...
                                                    'area', obj.gss_re.area, 'system', obj.pparams.sys);
                end
            end
               
            % Set up final phase encode gradients if needed     
            if obj.N_pe > 0
                obj.pe_grad1 = mr.makeTrapezoid(obj.pparams.channels{1}, 'duration', obj.refocus_time, ...
                                                        'area', obj.max_area, 'system', obj.pparams.sys);        
                
                obj.pe_grad2 = mr.makeTrapezoid(obj.pparams.channels{2}, 'duration', obj.refocus_time, ...
                                                        'area', obj.max_area, 'system', obj.pparams.sys);  
                
                obj.amp1 = obj.pe_grad1.amplitude;
                obj.amp2 = obj.pe_grad2.amplitude;
            end

        end

        function allblocks = build_blocks(obj, varargin)
            % Build all blocks to add to sequence.
            % 
            % Parameters
            % ----------
            % idx0, idx1 : int
            %     Index for the 0th channel area.  Will be ignored if the areas given for this
            %     channel were not inputted with multiple values.
            % 
            % offset: float
            %     Slice offset [m]
            % 
            % Returns
            % -------
            % cell array of cell arrays
            %     All blocks to add, outer list is for sequential blocks, inner blocks are all
            %     components to play within the block.

            parser = inputParser;
            addParameter(parser, 'idx1', 1);
            addParameter(parser, 'idx2', 1);
            addParameter(parser, 'seg_idx', -1);
            addParameter(parser, 'offset', 0);

            parse(parser, varargin{:});
            opt = parser.Results;

            obj.gss.channel = obj.pparams.channels{3};
            obj.rfp.freqOffset = obj.gss.amplitude * opt.offset;

            if obj.pparams.rf_spoil
                obj.rfp.phaseOffset = obj.pparams.rf_spoil_phase;
            end

            allblocks = {};

            % --- Slice select prewinder
            if obj.do_prewind
                obj.gss_re.channel = obj.pparams.channels{3};
                blocks = {obj.gss_re};
                if (opt.seg_idx >= 0)
                    blocks{end+1} = mr.makeLabel('SET', 'TRID', opt.seg_idx);
                end
                allblocks{end+1} = blocks;
            end

            % --- RF and slice select gradient
            blocks = {obj.rfp, obj.gss};
            if ((opt.seg_idx >= 0) && ~obj.do_prewind)
                blocks{end+1} = mr.makeLabel('SET', 'TRID', opt.seg_idx);
            end
            allblocks{end+1} = blocks;

            % --- Spatial encoding and slice select refocusing
            blocks = {};
            
            if obj.N_pe > 0
                area1 = obj.pe_areas(opt.idx1);
                area2 = obj.pe_areas(opt.idx2);
                
                obj.pe_grad1.channel = obj.pparams.channels{1};
                obj.pe_grad2.channel = obj.pparams.channels{2};
                
                obj.pe_grad1.amplitude = obj.amp1 * area1/obj.max_area;
                obj.pe_grad2.amplitude = obj.amp2 * area2/obj.max_area;
                
                blocks{end+1} = obj.pe_grad1;
                blocks{end+1} = obj.pe_grad2;
            end
                
                
            if obj.do_refocus
                obj.gss_re.channel = obj.pparams.channels{3};
                blocks{end+1} = obj.gss_re;
            end
                
            allblocks{end+1} = blocks;


        end

    end
end